File: player/apple2/src/combat.c
/* combat.c -- coeur portable du combat (cf. combat.h).
 *
 * Resolution v1 (livre-jeu) : chaque round, heros et ennemi lancent 2d6 + leur
 * attaque ; le plus haut score touche et inflige (degats - armure adverse),
 * minimum 1 (garantit la fin du combat). Les PV du heros sont la stat de PV
 * declaree (@combat_hp) ; l'arme/armure portee modifie att/degats/armure. */

#include "combat.h"
#include "story.h"
#include "state.h"

/* --- PRNG xorshift 16 bits --------------------------------------------- */
static u16 rng_state = 0xACE1;

void rng_seed(u16 s) { rng_state = s ? s : 0xACE1; }

static u16 rng_next(void)
{
    u16 x = rng_state;
    x ^= (u16)(x << 7);
    x ^= (u16)(x >> 9);
    x ^= (u16)(x << 8);
    rng_state = x;
    return x;
}

u8 rng_d6(void) { return (u8)(rng_next() % 6) + 1; }

/* --- Etat --------------------------------------------------------------- */
u16 cb_enemy_hp;
u8  cb_e_att, cb_e_dmg, cb_e_armor;
u8  cb_pscore, cb_escore, cb_last_dmg, cb_last_to;

void combat_begin(u8 att, u8 hp, u8 dmg, u8 armor)
{
    cb_e_att = att;
    cb_enemy_hp = hp;
    cb_e_dmg = dmg;
    cb_e_armor = armor;
    cb_pscore = cb_escore = cb_last_dmg = 0;
    cb_last_to = 2;
}

u8 combat_hero_hp(void)
{
    return (g_combat_hp == 0xFF) ? 0 : stat_val[g_combat_hp];
}

static void hero_take(u8 d)
{
    u8 h;
    if (g_combat_hp == 0xFF)
        return;
    h = stat_val[g_combat_hp];
    stat_val[g_combat_hp] = (d >= h) ? 0 : (u8)(h - d);
}

/* degats effectifs = max(1, base - armure) */
static u8 damage(int base, int armor)
{
    int d = base - armor;
    return (d < 1) ? 1 : (u8)d;
}

u8 combat_attack(void)
{
    int patt = (g_combat_att == 0xFF ? 0 : stat_val[g_combat_att]) + gear_atk();
    int pdmg = (int)g_combat_basedmg + gear_dmg();
    int parmor = gear_armor();
    int ps, es;

    if (patt < 0) patt = 0;
    ps = (int)rng_d6() + rng_d6() + patt;
    es = (int)rng_d6() + rng_d6() + cb_e_att;
    cb_pscore = (u8)(ps > 255 ? 255 : ps);
    cb_escore = (u8)(es > 255 ? 255 : es);

    if (ps > es) {                       /* le heros touche */
        u8 d = damage(pdmg, cb_e_armor);
        cb_enemy_hp = (d >= cb_enemy_hp) ? 0 : (u16)(cb_enemy_hp - d);
        cb_last_dmg = d;
        cb_last_to = 0;
    } else if (es > ps) {                /* l'ennemi touche */
        u8 d = damage(cb_e_dmg, parmor);
        hero_take(d);
        cb_last_dmg = d;
        cb_last_to = 1;
    } else {                            /* egalite : personne */
        cb_last_dmg = 0;
        cb_last_to = 2;
    }

    if (cb_enemy_hp == 0)
        return CB_VICTOIRE;
    if (combat_hero_hp() == 0)
        return CB_DEFAITE;
    return CB_CONTINUE;
}

u8 combat_flee(void)
{
    u8 d = damage(cb_e_dmg, gear_armor());   /* coup gratuit de l'ennemi */
    hero_take(d);
    cb_last_dmg = d;
    cb_last_to = 1;
    return (combat_hero_hp() == 0) ? CB_DEFAITE : CB_FUITE;
}

File: player/webng/src/app/engine/combat.ts
// combat.ts -- coeur portable du combat (RNG + resolution d'un round),
// miroir de combat.c. Meme regles v1 : chaque round, heros et ennemi
// lancent 2d6 + leur attaque ; le plus haut score touche et inflige
// max(1, degats - armure adverse).

import type { GameState } from "./state";
import type { StoryData } from "./story";
import { CB_CONTINUE, CB_FLEE, CB_LOSE, CB_WIN } from "./format";

/** xorshift 16 bits -- PRNG volontairement simple et rapide, pas
 * cryptographique (cf. combat.c pour le meme choix et la meme mise en garde :
 * predictible si on connait l'etat, mais suffisant pour des jets de d6). */
export class Rng {
  private state: number;
  constructor(seed = 0xace1) {
    this.state = seed || 0xace1;
  }
  seed(s: number): void {
    this.state = s || 0xace1;
  }
  private next(): number {
    let x = this.state & 0xffff;
    x ^= (x << 7) & 0xffff;
    x ^= x >> 9;
    x ^= (x << 8) & 0xffff;
    this.state = x & 0xffff;
    return this.state;
  }
  d6(): number {
    return (this.next() % 6) + 1;
  }
}

export type CombatOutcome =
  | typeof CB_WIN
  | typeof CB_LOSE
  | typeof CB_FLEE
  | typeof CB_CONTINUE;

export interface CombatRoundInfo {
  pscore: number;
  escore: number;
  lastDmg: number;
  lastTo: 0 | 1 | 2; // 0 = ennemi touche, 1 = heros touche, 2 = aucun (egalite)
}

function damage(base: number, armor: number): number {
  const d = base - armor;
  return d < 1 ? 1 : d;
}

export class Combat {
  enemyHp = 0;
  private eAtt = 0;
  private eDmg = 0;
  private eArmor = 0;
  last: CombatRoundInfo = { pscore: 0, escore: 0, lastDmg: 0, lastTo: 2 };

  private readonly rng: Rng;
  private readonly state: GameState;
  private readonly story: StoryData;

  constructor(rng: Rng, state: GameState, story: StoryData) {
    this.rng = rng;
    this.state = state;
    this.story = story;
  }

  begin(att: number, hp: number, dmg: number, armor: number): void {
    this.eAtt = att;
    this.enemyHp = hp;
    this.eDmg = dmg;
    this.eArmor = armor;
    this.last = { pscore: 0, escore: 0, lastDmg: 0, lastTo: 2 };
  }

  heroHp(): number {
    return this.story.combatHp === 0xff ? 0 : this.state.statVal[this.story.combatHp];
  }

  private heroTake(d: number): void {
    if (this.story.combatHp === 0xff) return;
    const h = this.state.statVal[this.story.combatHp];
    this.state.statVal[this.story.combatHp] = d >= h ? 0 : h - d;
  }

  attack(): CombatOutcome {
    const patt = Math.max(
      0,
      (this.story.combatAtt === 0xff ? 0 : this.state.statVal[this.story.combatAtt]) +
        this.state.gearAtk(),
    );
    const pdmg = this.story.combatBaseDmg + this.state.gearDmg();
    const parmor = this.state.gearArmor();

    const ps = this.rng.d6() + this.rng.d6() + patt;
    const es = this.rng.d6() + this.rng.d6() + this.eAtt;
    this.last.pscore = Math.min(ps, 255);
    this.last.escore = Math.min(es, 255);

    if (ps > es) {
      const d = damage(pdmg, this.eArmor);
      this.enemyHp = d >= this.enemyHp ? 0 : this.enemyHp - d;
      this.last.lastDmg = d;
      this.last.lastTo = 0;
    } else if (es > ps) {
      const d = damage(this.eDmg, parmor);
      this.heroTake(d);
      this.last.lastDmg = d;
      this.last.lastTo = 1;
    } else {
      this.last.lastDmg = 0;
      this.last.lastTo = 2;
    }

    if (this.enemyHp === 0) return CB_WIN;
    if (this.heroHp() === 0) return CB_LOSE;
    return CB_CONTINUE;
  }

  flee(): CombatOutcome {
    const d = damage(this.eDmg, this.state.gearArmor());
    this.heroTake(d);
    this.last.lastDmg = d;
    this.last.lastTo = 1;
    return this.heroHp() === 0 ? CB_LOSE : CB_FLEE;
  }
}
